File: src/129. Sum Root to Leaf Numbers/main.go
// Solved by @kitanoyoru

package main

import "strconv"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathToLeaf(root *TreeNode, path string, ans *int) {
	if root == nil {
		return
	}

	path += strconv.Itoa(root.Val)

	if root.Left == nil && root.Right == nil {
		if val, err := strconv.Atoi(path); err == nil {
			*ans += val
		}
		return
	}

	pathToLeaf(root.Left, path, ans)
	pathToLeaf(root.Right, path, ans)
}

func sumNumbers(root *TreeNode) int {
	ans := 0
	pathToLeaf(root, "", &ans)

	return ans
}

File: src/129. Sum Root to Leaf Numbers/main.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut sum = 0;
        Self::helper(&root, 0, &mut sum);
        return sum;
    }

    fn helper(root: &Option<Rc<RefCell<TreeNode>>>, mut path: i32, sum: &mut i32) {
        if let Some(root) = root {
            let root = root.borrow();

            path = path * 10 + root.val;

            if root.left.is_none() && root.right.is_none() {
                *sum += path;
                return;
            }

            Self::helper(&root.left, path, sum);
            Self::helper(&root.right, path, sum);
        }
    }
}